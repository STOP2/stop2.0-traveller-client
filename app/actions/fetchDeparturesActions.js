import config from '../config/config'

const USE_LOCATION_RADIUS = true
const LOCATION_RADIUS = 500

export const REQUEST_DEPARTURES = 'REQUEST_DEPARTURES'
export const RECEIVE_DEPARTURES = 'RECEIVE_DEPARTURES'
export const REQUEST_ERROR = 'REQUEST_ERROR'

const API_ENDPOINT = '/stops'
const API_ENDPOINT_WITH_BEACONS = '/stops/beacons'

export let requestDepartures = function(latitude, longitude)
{
    return {
        type: REQUEST_DEPARTURES,
        isFetching: true,
        latitude: latitude,
        longitude: longitude
    }
}

export let requestDeparturesWithBeacons = function(major, minor)
{
    return {
        type: REQUEST_DEPARTURES,
        isFetching: true,
        major: major,
        minor: minor
    }
}

export let receiveDepartures = function(json)
{
    return {
        type: RECEIVE_DEPARTURES,
        isFetching: false,
        isReady: true,
        error: false,
        departures: json.stops
    }
}

export let requestError = function()
{
    return {
        type: REQUEST_ERROR,
        error: true
    }
}

export let fetchDepartures = function(latitude, longitude, withBeacons)
{
    let radiusString = ''

    if (USE_LOCATION_RADIUS) radiusString = '&rad=' + LOCATION_RADIUS

    return dispatch =>
    {
        let fetchString

        if (withBeacons)
        {
            // here latitude =  major and longtitude = minor
            dispatch(requestDeparturesWithBeacons(latitude, longitude))
            fetchString = config.API_URL + API_ENDPOINT_WITH_BEACONS + '?major=' + latitude + '&minor=' + longitude + radiusString
            console.log('REQUESTING DEPARTURES WITH BEACONS')
            console.log(latitude)
            console.log(fetchString)
        }
        else
        {
            dispatch(requestDepartures(latitude, longitude))
            fetchString = config.API_URL + API_ENDPOINT + '?lat=' + latitude + '&lon=' + longitude + radiusString
        }


        return fetch(fetchString,
            {
                method: 'GET',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            })
            .then(response =>
              {
                if (response.ok)
                {
                    return response.json()
                }
                else
                {
                    dispatch(requestError())
                }
            })
            .then(json => dispatch(receiveDepartures(json)))
            .catch(error => dispatch(requestError(error)))
    }
}
