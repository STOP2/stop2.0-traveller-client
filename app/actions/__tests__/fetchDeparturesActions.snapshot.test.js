import mockStore from 'redux-mock-store'

import { fetchDepartures } from '../fetchDeparturesActions'

const store = mockStore()

beforeEach(() =>
{
    store.clearActions()
})

describe('fetchDepartures actions', () =>
{
    it('should handle RECEIVE_DEPARTURES action', async() =>
    {
        const response = '{"stops": [{"stop": {"schedule": [{"arrival": 7, "destination": "Rautatientori", "line": "55", "route_id": "HSL:1055", "trip_id": "HSL:1055_20161003_La_2_1412", "vehicle_type": 3}, {"arrival": 21, "destination": "Rautatientori", "line": "55", "route_id": "HSL:1055", "trip_id": "HSL:1055_20161003_La_2_1427", "vehicle_type": 3}], "stop_code": "3597", "stop_name": "A.I. Virtasen aukio"}}]}'

        fetch.mockResponseSuccess(response)

        await store.dispatch(fetchDepartures(60.203978, 24.9633573))

        expect(store.getActions()).toMatchSnapshot()
    })
})