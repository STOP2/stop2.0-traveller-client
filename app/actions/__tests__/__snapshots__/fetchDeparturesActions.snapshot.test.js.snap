exports[`fetchDepartures actions should handle RECEIVE_DEPARTURES action with beacons disabled 1`] = `
Array [
  Object {
    "isFetching": true,
    "latitude": 60.203978,
    "longitude": 24.9633573,
    "type": "REQUEST_DEPARTURES",
  },
  Object {
    "departures": Array [
      Object {
        "stop": Object {
          "schedule": Array [
            Object {
              "arrival": 7,
              "destination": "Rautatientori",
              "line": "55",
              "route_id": "HSL:1055",
              "trip_id": "HSL:1055_20161003_La_2_1412",
              "vehicle_type": 3,
            },
            Object {
              "arrival": 21,
              "destination": "Rautatientori",
              "line": "55",
              "route_id": "HSL:1055",
              "trip_id": "HSL:1055_20161003_La_2_1427",
              "vehicle_type": 3,
            },
          ],
          "stop_code": "3597",
          "stop_name": "A.I. Virtasen aukio",
        },
      },
    ],
    "error": false,
    "isFetching": false,
    "isReady": true,
    "type": "RECEIVE_DEPARTURES",
  },
]
`;

exports[`fetchDepartures actions should handle RECEIVE_DEPARTURES action with beacons enabled 1`] = `
Array [
  Object {
    "isFetching": true,
    "major": 60.203978,
    "minor": 24.9633573,
    "type": "REQUEST_DEPARTURES",
  },
  Object {
    "departures": Array [
      Object {
        "stop": Object {
          "schedule": Array [
            Object {
              "arrival": 7,
              "destination": "Rautatientori",
              "line": "55",
              "route_id": "HSL:1055",
              "trip_id": "HSL:1055_20161003_La_2_1412",
              "vehicle_type": 3,
            },
            Object {
              "arrival": 21,
              "destination": "Rautatientori",
              "line": "55",
              "route_id": "HSL:1055",
              "trip_id": "HSL:1055_20161003_La_2_1427",
              "vehicle_type": 3,
            },
          ],
          "stop_code": "3597",
          "stop_name": "A.I. Virtasen aukio",
        },
      },
    ],
    "error": false,
    "isFetching": false,
    "isReady": true,
    "type": "RECEIVE_DEPARTURES",
  },
]
`;

exports[`fetchDepartures actions should handle REQUEST_ERROR action 1`] = `
Object {
  "error": true,
  "type": "REQUEST_ERROR",
}
`;
